plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,284 @@
+	if plugin.BaseURL[len(plugin.BaseURL)-1:] != "/" {
+		plugin.BaseURL += "/"
+	}
+
+	transport := http.DefaultTransport
+
+	plugin.Client = &http.Client{Transport: transport, Timeout: DefaultRequestTimeout}
+	plugin.Query = plugin.buildQuery()
+
+	return nil
+}
+
+func (plugin *T128GraphQL) Gather(acc telegraf.Accumulator) error {
+	timestamp := time.Now().Round(time.Second)
+
+	plugin.retrieveMetrics(acc, timestamp)
+
+	return nil
+}
+
+func (plugin *T128GraphQL) buildQuery() string {
+	//allow multiple inputs like names[ComboEast,ComboWest] ?
+	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
+	query := "query MyQuery{" + replacer.Replace(plugin.EntryPoint) + "{"
+
+	for _, element := range plugin.Fields {
+		query += "\n" + element
+	}
+	query = strings.TrimSpace(query)
+	for _, element := range plugin.Tags {
+		query += "\n" + element
+	}
+
+	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+
+	return query
+}
+
+func (plugin *T128GraphQL) retrieveMetrics(acc telegraf.Accumulator, timestamp time.Time) {
+	request, err := plugin.createRequest()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to create a request for query %s: %w", plugin.Query, err))
+		return
+	}
+
+	response, err := plugin.Client.Do(request)
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(response.Body)
+
+	if err != nil {
+		template := "query error for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		template := fmt.Sprintf("status code %d not OK for metric ", response.StatusCode) + "%s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//decode json
+	jsonParsed, err := gabs.ParseJSON(buf.Bytes())
+	if err != nil {
+		acc.AddError(fmt.Errorf("invalid json response for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	jsonEntryPoint := plugin.BuildJsonEntryPoint()
+	jsonObj, err := jsonParsed.JSONPointer(jsonEntryPoint)
+	if err != nil {
+		template := "unexpected response for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//update acc
+	jsonChildren, err := jsonObj.Children()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to iterate on response nodes for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	//TODO: allow nested fields/tags
+	for _, child := range jsonChildren {
+		node := child.Data().(map[string]interface{})
+		fields := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		//allow integer conversion?
+		//TODO: fully implement multiple fields
+		for fieldRenamed, fieldName := range plugin.Fields {
+			if isNil(node[fieldName]) {
+				acc.AddError(fmt.Errorf("found empty data for metric %s: field %s", plugin.CollectorName, fieldName))
+				continue
+			}
+			fields[fieldRenamed] = node[fieldName]
+
+			for tagRenamed, tagName := range plugin.Tags {
```

Comment:
Why is this loop nested inside the fields loop? I'd imagine this basically looping over the fields and then tags and populating the appropriate lists, no?

Reply:
I didn't finish implementing this because there's some discussion needed. So if someone configures two fields say `field-1` and `field-2` in a collector called `test-metric`, I should get two metrics per node right? One metric called `test-metric-field-1` and another called `test-metric-field-2`

Reply:
With telegraf lines, you can have as many fields as you desire. For compactness, I think you should get one metric that holds all of the fields. You would get multiple metrics per each combination of tags. For instance if node was the only tag, you should get a metric per node with each having as many fields as the user configures.

As for the name of the metric (which is also called the measurement name), I'd use the `collector-name`.

Reply:
- [ ] Fields and tags separate loops